## Week 3 of the Capstone Project for the IBM Professional Data Science Certificate - Coursera

##### Import necessary libraries

In [113]:
import pandas as pd
import numpy as np
import requests
from bs4 import BeautifulSoup
from geopy.geocoders import Nominatim 
import folium as f
print ('Pandas,Numpy & BeatufiulSoup have been imported!')

Pandas,Numpy & BeatufiulSoup have been imported!


Use the Notebook to build the code to scrape the following Wikipedia page, https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M, in order to obtain the data that is in the table of postal codes and to transform the data into a pandas dataframe like the one shown below:

In [114]:
sourceweb = requests.get('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M').text
soup = BeautifulSoup(sourceweb, 'html.parser')

##### Getting the required information from the website without any further editting.

In [115]:
postalCodeList = []
boroughList = []
neighborhoodList = []

for row in soup.find('table').find_all('tr'):
    cells = row.find_all('td')
    if(len(cells) > 0):
        postalCodeList.append(cells[0].text)
        boroughList.append(cells[1].text)
        neighborhoodList.append(cells[2].text.rstrip('\n'))

##### Fill the table with the list from the website - 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'

In [116]:
toronto_neighorhood = [('PostalCode', postalCodeList),
                      ('Borough', boroughList),
                      ('Neighborhood', neighborhoodList)]
toronto_df = pd.DataFrame.from_dict(dict(toronto_neighorhood))
toronto_df.head()

Borough      Neighborhood PostalCode
0      Not assigned      Not assigned        M1A
1      Not assigned      Not assigned        M2A
2        North York         Parkwoods        M3A
3        North York  Victoria Village        M4A
4  Downtown Toronto      Harbourfront        M5A

##### Remove cells with "Not assigned" under borough and grouping neighborhoods sharing the same postal code

In [117]:
toronto_df_dropna = toronto_df[toronto_df.Borough != 'Not assigned'].reset_index(drop=True)
toronto_df_grouped = toronto_df_dropna.groupby(['PostalCode','Borough'], as_index=False).agg(lambda x: ','.join(x))
toronto_df_grouped.head()

PostalCode      Borough                          Neighborhood
0        M1B  Scarborough                         Rouge,Malvern
1        M1C  Scarborough  Highland Creek,Rouge Hill,Port Union
2        M1E  Scarborough       Guildwood,Morningside,West Hill
3        M1G  Scarborough                                Woburn
4        M1H  Scarborough                             Cedarbrae

##### Making unassigned neighborhoods the same as their respective borough. 

In [118]:
na_neigh_rows = toronto_df_grouped.Neighborhood == 'Not assigned'
toronto_df_grouped.loc[na_neigh_rows, 'Neighborhood'] = toronto_df_grouped.loc[na_neigh_rows, 'Borough']

##### Printing the number of rows of the dataframe after cleaning the data.

In [119]:
toronto_df_final = toronto_df_grouped
toronto_df_final.shape

(103, 3)

## Part 2

##### Getting latitude and longitude from the website specified as source

In [201]:
source2 = "http://cocl.us/Geospatial_data/Geospatial_Coordinates.csv"
coor = pd.read_csv(source2)
coor.columns = ['PostalCode', 'Latitude', 'Longitude']
coor.rename({'Postal Code':'PostalCode'},axis=1, inplace=True)
coor.head(15)

PostalCode   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476
5         M1J  43.744734 -79.239476
6         M1K  43.727929 -79.262029
7         M1L  43.711112 -79.284577
8         M1M  43.716316 -79.239476
9         M1N  43.692657 -79.264848
10        M1P  43.757410 -79.273304
11        M1R  43.750072 -79.295849
12        M1S  43.794200 -79.262029
13        M1T  43.781638 -79.304302
14        M1V  43.815252 -79.284577

##### Creating a new dataframe by merging coor and toronto_df_final with a final dataframe of 5 (PostalCode, Borough, Neighborhood, Latitude and Longitude) columns and 103 rows 

In [179]:
df_toronto2 = pd.merge(toronto_df_final,coor,how='left', on='PostalCode')
df_toronto2.head()

PostalCode      Borough                          Neighborhood   Latitude  \
0        M1B  Scarborough                         Rouge,Malvern  43.806686   
1        M1C  Scarborough  Highland Creek,Rouge Hill,Port Union  43.784535   
2        M1E  Scarborough       Guildwood,Morningside,West Hill  43.763573   
3        M1G  Scarborough                                Woburn  43.770992   
4        M1H  Scarborough                             Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

### Part 3

In [122]:
geolocator = Nominatim()
loc = geolocator.geocode('Toronto')
lat = loc.latitude
long = loc.longitude
print('The coordinates of the city of Toronto are Lat:{}, Long:{}.'.format(lat, long))

The coordinates of the city of Toronto are Lat:43.653963, Long:-79.387207.


In [123]:
plot_map = f.Map(location=[lat, long], zoom_start=9)
for lt, lng, label in zip(df_toronto2['Latitude'], df_toronto2['Longitude'], df_toronto2['Neighborhood']):
    label = f.Popup(label, parse_html=True)
    f.CircleMarker(
        [lt, lng],
        radius=8,
        popup=label,
        color='green',
        fill=True,
        fill_opacity=0.4,).add_to(plot_map) 
plot_map

#### Map is too hard to understand so I zoom in more and change parameters to make it easier to the eye.

In [124]:
plot_map = f.Map(location=[lat, long], zoom_start=12)
for lt, lng, label in zip(df_toronto2['Latitude'], df_toronto2['Longitude'], df_toronto2['Neighborhood']):
    label = f.Popup(label, parse_html=True)
    f.CircleMarker(
        [lt, lng],
        radius=10,
        popup=label,
        color='green',
        fill=True,
        fill_opacity=0.4,).add_to(plot_map) 
plot_map

##### We can see now all the different boroughs and their respective neighborhoods marked using their respective Latitute and Longitude.

![Toronto Map](/Map%20Toronto.JPG)

In [216]:
df_toronto3 = df_toronto2[df_toronto2['Borough'].str.contains("Toronto")]

In [154]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Fetching package metadata .............
Solving package specifications: .

# All requested packages already installed.
# packages in environment at /opt/conda/envs/DSX-Python35:
#
geopy                     1.18.1                     py_0    conda-forge
Fetching package metadata .............
Solving package specifications: .

# All requested packages already installed.
# packages in environment at /opt/conda/envs/DSX-Python35:
#
folium                    0.5.0                      py_0    conda-forge
Libraries imported.


In [217]:
df_toronto3.reset_index(drop=True, inplace=True)
df_toronto3

PostalCode           Borough  \
0         M4E      East Toronto   
1         M4K      East Toronto   
2         M4L      East Toronto   
3         M4M      East Toronto   
4         M4N   Central Toronto   
5         M4P   Central Toronto   
6         M4R   Central Toronto   
7         M4S   Central Toronto   
8         M4T   Central Toronto   
9         M4V   Central Toronto   
10        M4W  Downtown Toronto   
11        M4X  Downtown Toronto   
12        M4Y  Downtown Toronto   
13        M5A  Downtown Toronto   
14        M5B  Downtown Toronto   
15        M5C  Downtown Toronto   
16        M5E  Downtown Toronto   
17        M5G  Downtown Toronto   
18        M5H  Downtown Toronto   
19        M5J  Downtown Toronto   
20        M5K  Downtown Toronto   
21        M5L  Downtown Toronto   
22        M5N   Central Toronto   
23        M5P   Central Toronto   
24        M5R   Central Toronto   
25        M5S  Downtown Toronto   
26        M5T  Downtown Toronto   
27        M5V  Downtown Toronto   
28        M5W  Downtown Toronto   
29        M5X  Downtown Toronto   
30        M6G  Downtown Toronto   
31        M6H      West Toronto   
32        M6J      West Toronto   
33        M6K      West Toronto   
34        M6P      West Toronto   
35        M6R      West Toronto   
36        M6S      West Toronto   
37        M7Y      East Toronto   

                                         Neighborhood   Latitude  Longitude  
0                                         The Beaches  43.676357 -79.293031  
1                         The Danforth West,Riverdale  43.679557 -79.352188  
2                       The Beaches West,India Bazaar  43.668999 -79.315572  
3                                     Studio District  43.659526 -79.340923  
4                                       Lawrence Park  43.728020 -79.388790  
5                                    Davisville North  43.712751 -79.390197  
6                                  North Toronto West  43.715383 -79.405678  
7                                          Davisville  43.704324 -79.388790  
8                          Moore Park,Summerhill East  43.689574 -79.383160  
9   Deer Park,Forest Hill SE,Rathnelly,South Hill,...  43.686412 -79.400049  
10                                           Rosedale  43.679563 -79.377529  
11                         Cabbagetown,St. James Town  43.667967 -79.367675  
12                               Church and Wellesley  43.665860 -79.383160  
13                           Harbourfront,Regent Park  43.654260 -79.360636  
14                            Ryerson,Garden District  43.657162 -79.378937  
15                                     St. James Town  43.651494 -79.375418  
16                                        Berczy Park  43.644771 -79.373306  
17                                 Central Bay Street  43.657952 -79.387383  
18                             Adelaide,King,Richmond  43.650571 -79.384568  
19    Harbourfront East,Toronto Islands,Union Station  43.640816 -79.381752  
20            Design Exchange,Toronto Dominion Centre  43.647177 -79.381576  
21                      Commerce Court,Victoria Hotel  43.648198 -79.379817  
22                                           Roselawn  43.711695 -79.416936  
23                 Forest Hill North,Forest Hill West  43.696948 -79.411307  
24                  The Annex,North Midtown,Yorkville  43.672710 -79.405678  
25                      Harbord,University of Toronto  43.662696 -79.400049  
26            Chinatown,Grange Park,Kensington Market  43.653206 -79.400049  
27  CN Tower,Bathurst Quay,Island airport,Harbourf...  43.628947 -79.394420  
28                    Stn A PO Boxes 25 The Esplanade  43.646435 -79.374846  
29              First Canadian Place,Underground city  43.648429 -79.382280  
30                                           Christie  43.669542 -79.422564  
31                        Dovercourt Village,Dufferin  43.669005 -79.442259  
32                            Little Portugal,Trinity  43.64792

In [125]:
CLIENT_ID = 'GVLBHHQJK4K4IK0GRMXV3JAICI3K5Z4GUGEF02KX5QKVG5XX' # your Foursquare ID
CLIENT_SECRET = 'NCIODMMWHJN0VWWNBQFT132OVX55HVIYSWBYQOOUKHPXAWMT' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: GVLBHHQJK4K4IK0GRMXV3JAICI3K5Z4GUGEF02KX5QKVG5XX
CLIENT_SECRET:NCIODMMWHJN0VWWNBQFT132OVX55HVIYSWBYQOOUKHPXAWMT


In [218]:
neighborhood_latitude = df_toronto3.loc[0, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = df_toronto3.loc[0, 'Longitude'] # neighborhood longitude value

neighborhood_name = df_toronto3.loc[0, 'Neighborhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of The Beaches are 43.67635739999999, -79.2930312.


In [219]:
LIMIT = 100 # limit of number of venues returned by Foursquare API
radius = 700 # define radius

url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)
url 

'https://api.foursquare.com/v2/venues/explore?&client_id=GVLBHHQJK4K4IK0GRMXV3JAICI3K5Z4GUGEF02KX5QKVG5XX&client_secret=NCIODMMWHJN0VWWNBQFT132OVX55HVIYSWBYQOOUKHPXAWMT&v=20180605&ll=43.67635739999999,-79.2930312&radius=700&limit=100'

In [220]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5c3cbddddd5797602bcbe499'},
 'response': {'groups': [{'items': [{'reasons': {'count': 0,
       'items': [{'reasonName': 'globalInteractionReason',
         'summary': 'This spot is popular',
         'type': 'general'}]},
      'referralId': 'e-0-4f5a855be4b0a4baa1ae0063-0',
      'venue': {'categories': [{'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/food/vegetarian_',
          'suffix': '.png'},
         'id': '4bf58dd8d48988d1d3941735',
         'name': 'Vegetarian / Vegan Restaurant',
         'pluralName': 'Vegetarian / Vegan Restaurants',
         'primary': True,
         'shortName': 'Vegetarian / Vegan'}],
       'id': '4f5a855be4b0a4baa1ae0063',
       'location': {'address': '2188 Queen Street E',
        'cc': 'CA',
        'city': 'Toronto',
        'country': 'Canada',
        'crossStreet': 'Balsam Ave',
        'distance': 519,
        'formattedAddress': ['2188 Queen Street E (Balsam Ave)',
         'Toronto ON M431E6

In [221]:
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [222]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues = nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

name                     categories        lat        lng
0    Tori's Bakeshop  Vegetarian / Vegan Restaurant  43.672114 -79.290331
1  Beaches Bake Shop                         Bakery  43.680363 -79.289692
2    Ed's Real Scoop                 Ice Cream Shop  43.672630 -79.287993
3    The Fox Theatre            Indie Movie Theater  43.672801 -79.287272
4     The Beech Tree                      Gastropub  43.680493 -79.288846

In [223]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

43 venues were returned by Foursquare.


#### I have done the same process as in the Explore Manhattan lab - Following this I will repeat the same process for all neighborhoods in Toronto finding the respective nearby venues in every neighborhood. 

In [224]:
LIMIT = 100
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [225]:
Toronto_venues = getNearbyVenues(names=df_toronto3['Neighborhood'],
                                   latitudes=df_toronto3['Latitude'],
                                   longitudes=df_toronto3['Longitude']
                                  )

The Beaches
The Danforth West,Riverdale
The Beaches West,India Bazaar
Studio District
Lawrence Park
Davisville North
North Toronto West
Davisville
Moore Park,Summerhill East
Deer Park,Forest Hill SE,Rathnelly,South Hill,Summerhill West
Rosedale
Cabbagetown,St. James Town
Church and Wellesley
Harbourfront,Regent Park
Ryerson,Garden District
St. James Town
Berczy Park
Central Bay Street
Adelaide,King,Richmond
Harbourfront East,Toronto Islands,Union Station
Design Exchange,Toronto Dominion Centre
Commerce Court,Victoria Hotel
Roselawn
Forest Hill North,Forest Hill West
The Annex,North Midtown,Yorkville
Harbord,University of Toronto
Chinatown,Grange Park,Kensington Market
CN Tower,Bathurst Quay,Island airport,Harbourfront West,King and Spadina,Railway Lands,South Niagara
Stn A PO Boxes 25 The Esplanade
First Canadian Place,Underground city
Christie
Dovercourt Village,Dufferin
Little Portugal,Trinity
Brockton,Exhibition Place,Parkdale Village
High Park,The Junction South
Parkdale,Roncesvall

#### Let's see the size of the the new data frame Toronto_venues 

In [226]:
print(Toronto_venues.shape)
Toronto_venues.head()

(1708, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  The Beaches              43.676357              -79.293031   
1  The Beaches              43.676357              -79.293031   
2  The Beaches              43.676357              -79.293031   
3  The Beaches              43.676357              -79.293031   
4  The Beaches              43.676357              -79.293031   

                 Venue  Venue Latitude  Venue Longitude        Venue Category  
0  Grover Pub and Grub       43.679181       -79.297215                   Pub  
1            Starbucks       43.678798       -79.298045           Coffee Shop  
2        Upper Beaches       43.680563       -79.292869          Neighborhood  
3      Beaches Fitness       43.680319       -79.290991  Gym / Fitness Center  
4           Dip 'n Sip       43.678897       -79.297745           Coffee Shop

In [227]:
Toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Adelaide,King,Richmond                                                100   
Berczy Park                                                            54   
Brockton,Exhibition Place,Parkdale Village                             19   
Business Reply Mail Processing Centre 969 Eastern                      18   
CN Tower,Bathurst Quay,Island airport,Harbourfr...                     14   
Cabbagetown,St. James Town                                             47   
Central Bay Street                                                     82   
Chinatown,Grange Park,Kensington Market                               100   
Christie                                                               15   
Church and Wellesley                                                   87   
Commerce Court,Victoria Hotel                                         100   
Davisville                                                             36   
Davisville North                                                       10   
Deer Park,Forest Hill SE,Rathnelly,South Hill,S...                     14   
Design Exchange,Toronto Dominion Centre                               100   
Dovercourt Village,Dufferin                                            19   
First Canadian Place,Underground city                                 100   
Forest Hill North,Forest Hill West                                      4   
Harbord,University of Toronto                                          34   
Harbourfront East,Toronto Islands,Union Station                       100   
Harbourfront,Regent Park                                               49   
High Park,The Junction South                                           23   
Lawrence Park                                                           4   
Little Portugal,Trinity                                                66   
Moore Park,Summerhill East                                              4   
North Toronto West                                                     21   
Parkdale,Roncesvalles                                                  16   
Rosedale                                                                4   
Roselawn                                                                3   
Runnymede,Swansea                                                      40   
Ryerson,Garden District                                               100   
St. James Town                                                        100   
Stn A PO Boxes 25 The Esplanade                                        95   
Studio District                                                        39   
The Annex,North Midtown,Yorkville                                      24   
The Beaches                                                             5   
The Beaches West,India Bazaar                                          20   
The Danforth West,Riverdale                                            42   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Adelaide,King,Richmond                                                 100   
Berczy Park                                                             54   
Brockton,Exhibition Place,Parkdale Village                              19   
Business Reply Mail Processing Centre 969 Eastern                       18   
CN Tower,Bathurst Quay,Island airport,Harbourfr...                      14   
Cabbagetown,St. James Town                                              47   
Central Bay Street                                                      82   
Chinatown,Grange Park,Kensington Market                                100   
Christie                                                                15   
Church and Wellesley                                                    87   
Commerce Court,Victoria Hotel      

#### Let's find out how many unique categories can be curated from all the returned venues

In [228]:
print('There are {} uniques categories.'.format(len(Toronto_venues['Venue Category'].unique())))

There are 239 uniques categories.


#### Analyze Each Neighborhood

In [229]:
# one hot encoding
toronto_onehot = pd.get_dummies(Toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = Toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Yoga Studio  Adult Boutique  Afghan Restaurant  Airport  \
0            0               0                  0        0   
1            0               0                  0        0   
2            0               0                  0        0   
3            0               0                  0        0   
4            0               0                  0        0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                   0             0               0                0   
1                   0             0               0                0   
2                   0             0               0                0   
3                   0             0               0                0   
4                   0             0               0                0   

   Airport Terminal  American Restaurant  Antique Shop  Aquarium  \
0                 0                    0             0         0   
1                 0                    0             0         0   
2                 0                    0             0         0   
3                 0                    0             0         0   
4                 0                    0             0         0   

   Arepa Restaurant  Art Gallery  Art Museum  Arts & Crafts Store  \
0                 0            0           0                    0   
1                 0            0           0                    0   
2                 0            0           0                    0   
3                 0            0           0                    0   
4                 0            0           0                    0   

   Asian Restaurant  Auto Workshop  BBQ Joint  Baby Store  Bagel Shop  Bakery  \
0                 0              0          0           0           0       0   
1                 0              0          0           0           0       0   
2                 0              0          0           0           0       0   
3                 0              0          0           0           0       0   
4                 0              0          0           0           0       0   

   Bank  Bar  Baseball Stadium  Basketball Stadium  Beach  Beer Bar  \
0     0    0                 0                   0      0         0   
1     0    0                 0                   0      0         0   
2     0    0                 0                   0      0         0   
3     0    0                 0                   0      0         0   
4     0    0                 0                   0      0         0   

   Beer Store  Belgian Restaurant  Bistro  Board Shop  Boat or Ferry  \
0           0                   0       0           0              0   
1           0                   0       0           0              0   
2           0                   0       0           0              0   
3           0                   0       0           0              0   
4           0                   0       0           0              0   

   Bookstore  Boutique  Brazilian Restaurant  Breakfast Spot  Brewery  \
0          0         0                     0               0        0   
1          0         0                     0               0        0   
2          0         0                     0               0        0   
3          0         0                     0               0        0   
4          0         0                     0               0        0   

   Bubble Tea Shop  Building  Burger Joint  Burrito Place  Bus Line  Butcher  \
0                0         0             0              0         0        0   
1                0         0             0              0         0        0   
2                0         0             0              0         0        0   
3                0         0             0              0         0        0   
4                0         0             0              0         0        0   

   Café  Cajun / Creole Restaurant  Caribbean Restaurant  Cheese Shop  \
0     0                          

##### Size of the dataframe

In [230]:
toronto_onehot.shape

(1708, 239)

#### Row Grouping by neighborhood and by taking the mean of the frequency of occurrence of each category

In [231]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood  Yoga Studio  \
0                              Adelaide,King,Richmond     0.000000   
1                                         Berczy Park     0.000000   
2          Brockton,Exhibition Place,Parkdale Village     0.000000   
3   Business Reply Mail Processing Centre 969 Eastern     0.055556   
4   CN Tower,Bathurst Quay,Island airport,Harbourf...     0.000000   
5                          Cabbagetown,St. James Town     0.000000   
6                                  Central Bay Street     0.012195   
7             Chinatown,Grange Park,Kensington Market     0.000000   
8                                            Christie     0.000000   
9                                Church and Wellesley     0.011494   
10                      Commerce Court,Victoria Hotel     0.000000   
11                                         Davisville     0.000000   
12                                   Davisville North     0.000000   
13  Deer Park,Forest Hill SE,Rathnelly,South Hill,...     0.000000   
14            Design Exchange,Toronto Dominion Centre     0.000000   
15                        Dovercourt Village,Dufferin     0.000000   
16              First Canadian Place,Underground city     0.000000   
17                 Forest Hill North,Forest Hill West     0.000000   
18                      Harbord,University of Toronto     0.000000   
19    Harbourfront East,Toronto Islands,Union Station     0.000000   
20                           Harbourfront,Regent Park     0.000000   
21                       High Park,The Junction South     0.000000   
22                                      Lawrence Park     0.000000   
23                            Little Portugal,Trinity     0.015152   
24                         Moore Park,Summerhill East     0.000000   
25                                 North Toronto West     0.047619   
26                              Parkdale,Roncesvalles     0.000000   
27                                           Rosedale     0.000000   
28                                           Roselawn     0.000000   
29                                  Runnymede,Swansea     0.000000   
30                            Ryerson,Garden District     0.000000   
31                                     St. James Town     0.010000   
32                    Stn A PO Boxes 25 The Esplanade     0.010526   
33                                    Studio District     0.025641   
34                  The Annex,North Midtown,Yorkville     0.000000   
35                                        The Beaches     0.000000   
36                      The Beaches West,India Bazaar     0.000000   
37                        The Danforth West,Riverdale     0.023810   

    Adult Boutique  Afghan Restaurant   Airport  Airport Food Court  \
0         0.000000           0.000000  0.000000            0.000000   
1         0.000000           0.000000  0.000000            0.000000   
2         0.000000           0.000000  0.000000            0.000000   
3         0.000000           0.000000  0.000000            0.000000   
4         0.000000           0.000000  0.071429            0.071429   
5         0.000000           0.000000  0.000000            0.000000   
6         0.000000           0.000000  0.000000            0.000000   
7         0.000000           0.000000  0.000000            0.000000   
8         0.000000           0.000000  0.000000            0.000000   
9         0.011494           0.011494  0.000000            0.000000   
10        0.000000           0.000000  0.000000            0.000000   
11        0.000000           0.000000  0.000000            0.000000   
12        0.000000           0.000000  0.000000            0.000000   
13        0.000000           0.000000  0.000000            0.000000   
14        0.000000           0.000000  0.000000            0.000000   
15        0.000000           0.000000  0.000000            0.000000   
16        0.000000           0.000000  0.000000            0.000000   
17        0.000000           0.0

#### The new shape of the grouped dataframe is:

In [232]:
toronto_grouped.shape

(38, 239)

#### Top 5 venues for each neighborhood in the City of Toronto.

In [233]:
num_top_venues = 5

for hood in toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Adelaide,King,Richmond----
                 venue  freq
0          Coffee Shop  0.06
1                 Café  0.05
2           Steakhouse  0.04
3      Thai Restaurant  0.04
4  American Restaurant  0.04


----Berczy Park----
            venue  freq
0     Coffee Shop  0.07
1      Restaurant  0.06
2    Cocktail Bar  0.06
3          Bakery  0.04
4  Farmers Market  0.04


----Brockton,Exhibition Place,Parkdale Village----
            venue  freq
0  Breakfast Spot  0.11
1            Café  0.11
2     Coffee Shop  0.11
3    Climbing Gym  0.05
4             Bar  0.05


----Business Reply Mail Processing Centre 969 Eastern----
           venue  freq
0    Yoga Studio  0.06
1  Burrito Place  0.06
2  Garden Center  0.06
3         Garden  0.06
4            Spa  0.06


----CN Tower,Bathurst Quay,Island airport,Harbourfront West,King and Spadina,Railway Lands,South Niagara----
              venue  freq
0    Airport Lounge  0.14
1   Airport Service  0.14
2  Airport Terminal  0.14
3  Sculpture Garden

In [234]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

#### Top 10 venues

In [235]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted

Neighborhood 1st Most Common Venue  \
0                              Adelaide,King,Richmond           Coffee Shop   
1                                         Berczy Park           Coffee Shop   
2          Brockton,Exhibition Place,Parkdale Village           Coffee Shop   
3   Business Reply Mail Processing Centre 969 Eastern           Yoga Studio   
4   CN Tower,Bathurst Quay,Island airport,Harbourf...        Airport Lounge   
5                          Cabbagetown,St. James Town           Coffee Shop   
6                                  Central Bay Street           Coffee Shop   
7             Chinatown,Grange Park,Kensington Market                   Bar   
8                                            Christie                  Café   
9                                Church and Wellesley   Japanese Restaurant   
10                      Commerce Court,Victoria Hotel           Coffee Shop   
11                                         Davisville           Pizza Place   
12                                   Davisville North                  Park   
13  Deer Park,Forest Hill SE,Rathnelly,South Hill,...                   Pub   
14            Design Exchange,Toronto Dominion Centre           Coffee Shop   
15                        Dovercourt Village,Dufferin              Pharmacy   
16              First Canadian Place,Underground city           Coffee Shop   
17                 Forest Hill North,Forest Hill West              Bus Line   
18                      Harbord,University of Toronto                  Café   
19    Harbourfront East,Toronto Islands,Union Station           Coffee Shop   
20                           Harbourfront,Regent Park           Coffee Shop   
21                       High Park,The Junction South    Mexican Restaurant   
22                                      Lawrence Park              Bus Line   
23                            Little Portugal,Trinity                   Bar   
24                         Moore Park,Summerhill East            Playground   
25                                 North Toronto West   Sporting Goods Shop   
26                              Parkdale,Roncesvalles        Breakfast Spot   
27                                           Rosedale                  Park   
28                                           Roselawn                  Pool   
29                                  Runnymede,Swansea           Coffee Shop   
30                            Ryerson,Garden District           Coffee Shop   
31                                     St. James Town           Coffee Shop   
32                    Stn A PO Boxes 25 The Esplanade           Coffee Shop   
33                                    Studio District                  Café   
34                  The Annex,North Midtown,Yorkville           Coffee Shop   
35                                        The Beaches           Coffee Shop   
36                      The Beaches West,India Bazaar                  Park   
37                        The Danforth West,Riverdale      Greek Restaurant   

   2nd Most Common Venue          3rd Most Common Venue 4th Most Common Venue  \
0                   Café                     Steakhouse   American Restaurant   
1             Restaurant                   Cocktail Bar                  Café   
2                   Café                 Breakfast Spot                   Bar   
3          Auto Workshop                           Park           Pizza Place   
4        Airport Service               Airport Terminal                 Plane   
5             Restaurant             Italian Restaurant           Pizza Place   
6                   Café             Italian Restaurant          Burger Joint   
7                   Café  Vegetarian / Vegan Restaurant           Coffee Shop   
8          Grocery Store                           Park            Baby Store   
9            Coffee Shop               Sushi Restaurant               Gay Bar   
10                  Café                     Restaurant             

##### Running K means to cluster neighborhoods.

In [236]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([2, 2, 2, 1, 2, 2, 2, 2, 2, 2], dtype=int32)

In [237]:
toronto_merged = df_toronto3

# add clustering labels
toronto_merged['Cluster Labels'] = kmeans.labels_

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head() # check the last columns!

/opt/conda/envs/DSX-Python35/lib/python3.5/site-packages/ipykernel/__main__.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy


PostalCode          Borough                   Neighborhood   Latitude  \
0        M4E     East Toronto                    The Beaches  43.676357   
1        M4K     East Toronto    The Danforth West,Riverdale  43.679557   
2        M4L     East Toronto  The Beaches West,India Bazaar  43.668999   
3        M4M     East Toronto                Studio District  43.659526   
4        M4N  Central Toronto                  Lawrence Park  43.728020   

   Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0 -79.293031               2           Coffee Shop  Gym / Fitness Center   
1 -79.352188               2      Greek Restaurant           Coffee Shop   
2 -79.315572               2                  Park           Pizza Place   
3 -79.340923               1                  Café           Coffee Shop   
4 -79.388790               2              Bus Line                  Park   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0                   Pub         Women's Store        Discount Store   
1        Ice Cream Shop             Bookstore    Italian Restaurant   
2                   Pub  Fast Food Restaurant            Board Shop   
3                Bakery    Italian Restaurant   American Restaurant   
4    Dim Sum Restaurant           Swim School         Women's Store   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0  Fast Food Restaurant        Farmers Market    Falafel Restaurant   
1           Yoga Studio        Cosmetics Shop               Brewery   
2          Liquor Store     Fish & Chips Shop          Burger Joint   
3           Yoga Studio       Coworking Space    Seafood Restaurant   
4   Filipino Restaurant  Fast Food Restaurant        Farmers Market   

  9th Most Common Venue 10th Most Common Venue  
0           Event Space   Ethiopian Restaurant  
1       Bubble Tea Shop             Restaurant  
2        Sandwich Place          Burrito Place  
3        Sandwich Place            Cheese Shop  
4    Falafel Restaurant            Event Space

In [240]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=13)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i+x+(i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'],toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

#### Examining each cluster

### Cluster 1

In [241]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
24   Central Toronto               0           Coffee Shop   
27  Downtown Toronto               0        Airport Lounge   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
24                  Café        Sandwich Place           Pizza Place   
27       Airport Service      Airport Terminal                 Plane   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
24        History Museum     Indian Restaurant              Pharmacy   
27              Boutique       Harbor / Marina               Airport   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
24        Cosmetics Shop                   Pub           Burger Joint  
27    Airport Food Court          Airport Gate          Boat or Ferry

# Cluster 2

In [242]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
3       East Toronto               1                  Café   
12  Downtown Toronto               1   Japanese Restaurant   
22   Central Toronto               1                  Pool   
36      West Toronto               1           Coffee Shop   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
3            Coffee Shop                Bakery    Italian Restaurant   
12           Coffee Shop      Sushi Restaurant               Gay Bar   
22           Music Venue                Garden         Women's Store   
36           Pizza Place                  Café      Sushi Restaurant   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
3    American Restaurant           Yoga Studio       Coworking Space   
12            Restaurant          Burger Joint                  Café   
22               Dog Run  Fast Food Restaurant        Farmers Market   
36                 Diner    Italian Restaurant                   Gym   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
3     Seafood Restaurant        Sandwich Place            Cheese Shop  
12                   Pub       Bubble Tea Shop            Men's Store  
22    Falafel Restaurant           Event Space   Ethiopian Restaurant  
36             Bookstore                  Food                    Bar

# Cluster 3

In [243]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0       East Toronto               2           Coffee Shop   
1       East Toronto               2      Greek Restaurant   
2       East Toronto               2                  Park   
4    Central Toronto               2              Bus Line   
5    Central Toronto               2                  Park   
6    Central Toronto               2   Sporting Goods Shop   
7    Central Toronto               2           Pizza Place   
8    Central Toronto               2            Playground   
9    Central Toronto               2                   Pub   
10  Downtown Toronto               2                  Park   
11  Downtown Toronto               2           Coffee Shop   
13  Downtown Toronto               2           Coffee Shop   
14  Downtown Toronto               2           Coffee Shop   
15  Downtown Toronto               2           Coffee Shop   
16  Downtown Toronto               2           Coffee Shop   
18  Downtown Toronto               2           Coffee Shop   
19  Downtown Toronto               2           Coffee Shop   
20  Downtown Toronto               2           Coffee Shop   
21  Downtown Toronto               2           Coffee Shop   
23   Central Toronto               2              Bus Line   
25  Downtown Toronto               2                  Café   
26  Downtown Toronto               2                   Bar   
29  Downtown Toronto               2           Coffee Shop   
30  Downtown Toronto               2                  Café   
31      West Toronto               2              Pharmacy   
32      West Toronto               2                   Bar   
33      West Toronto               2           Coffee Shop   
34      West Toronto               2    Mexican Restaurant   
35      West Toronto               2        Breakfast Spot   
37      East Toronto               2           Yoga Studio   

   2nd Most Common Venue          3rd Most Common Venue 4th Most Common Venue  \
0   Gym / Fitness Center                            Pub         Women's Store   
1            Coffee Shop                 Ice Cream Shop             Bookstore   
2            Pizza Place                            Pub  Fast Food Restaurant   
4                   Park             Dim Sum Restaurant           Swim School   
5         Breakfast Spot                   Burger Joint     Food & Drink Shop   
6            Coffee Shop                 Clothing Store    Salon / Barbershop   
7           Dessert Shop                 Sandwich Place    Italian Restaurant   
8                    Gym                   Tennis Court                  Park   
9            Coffee Shop            American Restaurant    Light Rail Station   
10            Playground                          Trail                 Diner   
11            Restaurant             Italian Restaurant           Pizza Place   
13                Bakery                            Pub                  Park   
14        Clothing Store                           Café        Cosmetics Shop   
15            Restaurant                           Café                 Hotel   
16            Restaurant                   Cocktail Bar                  Café   
18                  Café                     Steakhouse   American Restaurant   
19              Aquarium                          Hotel                  Café   
20                 Hotel                           Café   American Restaurant   
21                  Café                     Restaurant                 Hotel   
23                 Trail               Sushi Restaurant         Jewelry Store   
25           Coffee Shop                      Bookstore            Restaurant   
26                  Café  Vegetarian / Vegan Restaurant           Coffee Shop   
29                  Café                          Hotel            Restaurant   
30         Grocery Store                           Park            Baby Store   
31           Supermarket                 Discount Store          

# Cluster 4

In [244]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
17  Downtown Toronto               3           Coffee Shop   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
17                  Café    Italian Restaurant          Burger Joint   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
17                   Bar       Thai Restaurant    Falafel Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
17       Bubble Tea Shop    Chinese Restaurant      Indian Restaurant

# Cluster 5

In [245]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
28  Downtown Toronto               4           Coffee Shop   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
28            Restaurant                  Café    Italian Restaurant   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
28              Beer Bar    Seafood Restaurant                   Pub   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
28                 Hotel          Cocktail Bar                 Bakery